src/java.base/share/classes/java/security/PEMEncoder.java

```diff
@@ -0,0 +1,330 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.security;
+
+import jdk.internal.javac.PreviewFeature;
+
+import sun.security.pkcs.PKCS8Key;
+import sun.security.util.DerOutputStream;
+import sun.security.util.DerValue;
+import sun.security.util.Pem;
+import sun.security.x509.AlgorithmId;
+
+import javax.crypto.*;
+import javax.crypto.spec.PBEKeySpec;
+import java.io.*;
+import java.nio.charset.StandardCharsets;
+import java.security.cert.*;
+import java.security.cert.Certificate;
+import java.security.spec.AlgorithmParameterSpec;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.Objects;
+
+/**
+ * PEMEncoder is an immutable Privacy-Enhanced Mail (PEM) encoding class.
+ * PEM is a textual encoding used for storing and transferring security
+ * objects, such as asymmetric keys, certificates, and certificate revocation
+ * lists (CRL). Defined in RFC 1421 and RFC 7468, PEM consists of a
+ * Base64-formatted binary encoding surrounded by a type identifying header
+ * and footer.
+ * <p>
+ * Encoding may be performed on objects that implement {@link DEREncodable}.
+ * <p>
+ * Encrypted private key PEM data can be built by calling the encode methods
+ * on a PEMEncoder instance returned by {@link #withEncryption(char[])} or
+ * by passing an {@link EncryptedPrivateKeyInfo} object into the encode methods.
+ * <p>
+ * PKCS8 v2.0 allows OneAsymmetric encoding, which is a private and public
+ * key in the same PEM.  This is supported by using the {@link KeyPair} class
+ * with the encode methods.
+ * <p>
+ * PEMEncoder supports the follow types:
+ * <pre>
+ *     PRIVATE KEY, PUBLIC KEY, CERTIFICATE, CRL, and ENCRYPTED PRIVATE KEY.
+ * </pre>
+ *
+ * @apiNote
+ * Here is an example of encoding a PrivateKey object:
+ * <pre>
+ *     PEMEncoder pe = PEMEncoder.of();
+ *     byte[] pemData = pe.encode(privKey);
+ * </pre>
+ *
+ * @since 24
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.PEM_API)
+public final class PEMEncoder {
+
+    // Singleton instance of PEMEncoder
+    private static final PEMEncoder PEM_ENCODER = new PEMEncoder(null);
+
+    // If non-null, encoder is configured for encryption
+    private Cipher cipher = null;
```

Comment:
if cipher is defined this is an encrypted PEMEncoder instance, so yes it's important.
Encoding doesn't need a factory when the object already provides the DER encoding.

Reply:
Not sure if I understand. When using encryption, user can set a cipher algorithm that is not implemented in any of the builtin providers, so the `SecretKeyFactory.getInstance` and `Cipher.getInstance` might need a provider argument.

Reply:
What you are describing is why the encoder accepts EncryptedPrivateKeyInfo.  These more complicated configurations I left to EncryptedPrivateKeyInfo as the PEM classes are not trying to replace all the functionality of EncryptedPrivateKeyInfo or the other classes, just simplify the common cases and the PEM construction.
PEMDecoder has a factory method because I saw that as a more common case for important into a HSM.

Reply:
I see.

Reply:
I think I see what you were first saying.  Cipher doesn't need to be a class variable.